testutils/src/main/java/org/zaproxy/zap/testutils/ScanRuleTests.java

```diff
@@ -0,0 +1,95 @@
+/*
+ * Zed Attack Proxy (ZAP) and its related class files.
+ *
+ * ZAP is an HTTP/HTTPS proxy for assessing web application security.
+ *
+ * Copyright 2023 The ZAP Development Team
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.zaproxy.zap.testutils;
+
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.empty;
+import static org.hamcrest.Matchers.is;
+import static org.junit.jupiter.api.DynamicTest.dynamicTest;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+import org.apache.commons.httpclient.URI;
+import org.junit.jupiter.api.DynamicTest;
+import org.junit.jupiter.api.TestFactory;
+import org.parosproxy.paros.core.scanner.Alert;
+import org.parosproxy.paros.network.HttpHeader;
+import org.zaproxy.zap.extension.alert.ExampleAlertProvider;
+
+interface ScanRuleTests {
+
+    Object getScanRule();
+
+    @TestFactory
+    default Collection<DynamicTest> addScanRuleTests() {
+        List<DynamicTest> tests = new ArrayList<>();
+        if (getScanRule() instanceof ExampleAlertProvider) {
+            tests.add(dynamicTest("shouldHaveValidReferences", this::shouldHaveValidReferences));
+        }
+        return tests;
+    }
+
+    private void shouldHaveValidReferences() {
+        // Given / When
+        List<Alert> alerts = ((ExampleAlertProvider) getScanRule()).getExampleAlerts();
+        // Then
+        if (alerts == null || alerts.isEmpty()) {
+            return;
+        }
+
+        Set<String> references =
+                alerts.stream()
+                        .map(Alert::getReference)
+                        .map(refs -> refs.split("\\r?\\n"))
+                        .flatMap(Arrays::stream)
+                        .filter(e -> !e.isBlank())
+                        .filter(e -> e.startsWith(HttpHeader.SCHEME_HTTP))
+                        .map(
+                                e -> {
+                                    int idx = e.indexOf(' ');
```

Comment:
Handles references with appended text, should we allow this?
e.g.: `nhttp://www.w3.org/Protocols/rfc2616/rfc2616-sec13.html (obsoleted by rfc7234)` (the site has wrong URL https://www.zaproxy.org/docs/alerts/10049-1/ )

Reply:
Some of these were already cleaned up and the validations will disallow extra text.